### PR TITLE
Split compute and display gitlab ci config / diff ACIX-423

### PR DESCRIPTION
**Pull request by CelianR**
### What does this PR do?

Preparing ACIX-423

Currently, only the gitlab diff commenter is using the diff but soon more jobs such as linters will use it as well (cf ACIX-423).
Split into two stages the job as computing the diff is costly.

Added serialization functions and an `iter_jobs` method to iterate easily through all jobs of a diff.

> [!NOTE]
> I used arm64 runners as recommended for the new / modified jobs

### CODE REVIEW

File: .gitlab/notify/notify.yml [hunk 1]
```diff
@@ -91,7 +91,8 @@ notify_github:
 notify_gitlab_ci_changes:
   image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-buildimages/deb_x64$DATADOG_AGENT_BUILDIMAGES_SUFFIX:$DATADOG_AGENT_BUILDIMAGES
   stage: notify
-  needs: []
+  needs: [compute_gitlab_ci_config]
+  dependencies: [compute_gitlab_ci_config]
```
Comment (chouetz) [inline .gitlab/notify/notify.yml:95]: If I'm correct the artifact of the `needs` job list is [already available by default](https://docs.gitlab.com/ee/ci/yaml/index.html#needsartifacts)
```suggestion
  needs: [compute_gitlab_ci_config]
```

File: tasks/gitlab_helpers.py [hunk 2]
```diff
@@ -281,3 +283,57 @@ def print_entry_points(ctx):
     print(len(entry_points), 'entry points:')
     for entry_point, config in entry_points.items():
         print(f'- {color_message(entry_point, Color.BOLD)} ({len(config)} components)')
+
+
+def compute_gitlab_ci_config_diff(ctx, before: str, after: str):
+    """
+    Computes the full configs and the diff between two git references.
+    The base commit where after is compared to is the LCA commit between before and after.
```
Comment (chouetz) [inline tasks/gitlab_helpers.py:291]:
```suggestion
    The "after reference" is compared to the Lowest Common Ancestor (LCA) commit of "before reference" and "after reference".
```
Don't hesitate to rephrase:
- I think it's nice to put the meaning of an acronym before using it
- using quotes help reading the sentence

File: tasks/gitlab_helpers.py [hunk 3]
```diff
@@ -281,3 +283,57 @@ def print_entry_points(ctx):
...
+    """
+
+    before_name = before or "merge base"
+    after_name = after or "local files"
```
Comment (chouetz) [inline tasks/gitlab_helpers.py:295]:
```suggestion
    after_name = after or "HEAD"
```
I suppose. We could also use [this](https://github.com/DataDog/datadog-agent/blob/main/tasks/libs/common/git.py#L53)

File: tasks/gitlab_helpers.py [hunk 4]
```diff
@@ -281,3 +283,57 @@ def print_entry_points(ctx):
...
+
+    # The before commit is the LCA commit between before and after
+    before = before or DEFAULT_BRANCH
+    before = ctx.run(f'git merge-base {before} {after or "HEAD"}', hide=True).stdout.strip()
```
Comment (chouetz) [inline tasks/gitlab_helpers.py:299]: We can use [this](https://github.com/DataDog/datadog-agent/blob/main/tasks/libs/common/git.py#L57) I guess

File: tasks/gitlab_helpers.py [hunk 5]
```diff
@@ -281,3 +283,57 @@ def print_entry_points(ctx):
     print(len(entry_points), 'entry points:')
     for entry_point, config in entry_points.items():
         print(f'- {color_message(entry_point, Color.BOLD)} ({len(config)} components)')
+
+
+def compute_gitlab_ci_config_diff(ctx, before: str, after: str):
```
Comment (chouetz) [inline tasks/gitlab_helpers.py:288]: I think we should try to keep only task definitions on modules under the `tasks` folder. Can we put this method on a dedicated `lib`?

File: tasks/gitlab_helpers.py [hunk 6]
```diff
@@ -281,3 +283,57 @@ def print_entry_points(ctx):
...
+    before: str | None = None,
+    after: str | None = None,
+    before_file: str = 'before.gitlab-ci.yml',
+    after_file: str = 'after.gitlab-ci.yml',
+    diff_file: str = 'diff.gitlab-ci.yml',
+):
+    """
+    Will compute the Gitlab CI full configuration for the current commit and the base commit and will compute the diff between them.
+
+    The diff can be loaded using `MultiGitlabCIDiff.from_dict`.
```
Comment (chouetz) [inline tasks/gitlab_helpers.py:324]: 🥜 I think this is not necessary

File: tasks/libs/ciproviders/gitlab_api.py [hunk 7]
```diff
@@ -116,20 +116,48 @@ def refresh_pipeline(pipeline: ProjectPipeline):
...
+        self, before: dict | None = None, after: dict | None = None, dict_initialization: dict | None = None
+    ) -> None:
         """
         Used to display job diffs between two gitlab ci configurations
         """
-        self.before = before
-        self.after = after
-        self.added_contents = {}
-        self.modified_diffs = {}
+        if dict_initialization:
```
Comment (chouetz) [inline tasks/libs/ciproviders/gitlab_api.py:125]: 💬 I am wondering if having `GitlabCIDiff` inherit from a [UserDict](https://docs.python.org/3/library/collections.html#collections.UserDict) could make it more natural

File: tasks/libs/ciproviders/gitlab_api.py [hunk 8]
```diff
@@ -336,6 +364,27 @@ def str_note() -> list[str]:
 
         return '\n'.join(res)
 
+    def iter_jobs(self, added=True, modified=True, removed=False):
+        """
+        Will iterate over all jobs in all files for the given states
+
+        Returns a tuple of (job_name, contents, state)
+
+        Note that the contents is the contents after modification or before removal
```
Comment (chouetz) [inline tasks/libs/ciproviders/gitlab_api.py:373]: This is not really clear, could you clarify?

**Review by chouetz — APPROVED** [review]
Minor comments, nice job

File: tasks/gitlab_helpers.py (hunk 3 above)
Comment (CelianR) [reply tasks/gitlab_helpers.py:295]: By default, this is not always `HEAD`. If you test locally and you have uncommitted changes, then local files will be parsed and not files from `HEAD`

File: tasks/libs/ciproviders/gitlab_api.py (hunk 8 above)
Comment (CelianR) [reply tasks/libs/ciproviders/gitlab_api.py:389]: > Note that the contents of the job is the contents after modification if modified or before removal if removed

**Pull request merged**